### PR TITLE
Enable loading CuDNNLSTM weights into an LSTM layer

### PR DESCRIPTION
**Pull request by myutwo150**
This PR fixes a `ValueError` when trying to load weights from a `CuDNNLSTM` layer into an `LSTM` layer. Because `CuDNNLSTM` uses two biases, the size of `self.bias` is doubled comparing to that of an `LSTM` layer.

From the equations on the cuDNN user guide, by summing the two biases into one, the resulting bias term will be the same one used by `LSTM`.
screenshot 2017-10-31 2 09 17

However, there's no such relationship between `CuDNNGRU` and `GRU` (see the third equation in the image below, also from the cuDNN user guide). So this PR only deals with loading `CuDNNLSTM` weights.
screenshot 2017-10-31 2 11 38

Changes:

1. In `preprocess_weights_for_loading()`, check if we're loading `CuDNNLSTM` weights into `LSTM` by checking the shape of bias. If the size of bias is twice as large as `kernel.shape[1]`, then the weights are from an `CuDNNLSTM` layer.
2. Fix an issue in `_canonical_to_params()` where the kernel should be transposed before getting flattened.
I couldn't find it on the cuDNN user guide, but without transposing, the predictions from the two layers will be quite different. The kernel is also transposed [in TF](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py#L393-L403) before being fed into `_canonical_to_params()`.
3. Add a test to verify that `CuDNNLSTM` gives the same prediction as `LSTM` if the weights are the same.

### CODE REVIEW

File: keras/layers/cudnn_recurrent.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def __init__(self,
 
     def _canonical_to_params(self, weights, biases):
         import tensorflow as tf
-        weights = [tf.reshape(x, (-1,)) for x in weights]
+        weights = [tf.reshape(tf.transpose(x), (-1,)) for x in weights]
```
Comment (fchollet) [inline keras/layers/cudnn_recurrent.py]: I think this would break existing checkpoints. Have you tried?

Comment (myutwo150) [reply keras/layers/cudnn_recurrent.py]: Yes it's a breaking change. If it's important not to break existing checkpoints, I can reshape the kernel during weight loading. For example, adding these lines in `preprocess_weights_for_loading()` will also work:
```python
# reshape the kernels
kernels = np.split(weights[0], 4, axis=1)
kernels = [kernel.reshape(-1).reshape(kernel.shape, order='F') for kernel in kernels]
weights[0] = np.concatenate(kernels, axis=1)

# reshape the recurrent kernels
recurrent_kernels = np.split(weights[1], 4, axis=1)
recurrent_kernels = [kernel.T for kernel in recurrent_kernels]
weights[1] = np.concatenate(recurrent_kernels, axis=1)
```
But if we fix it like this, the kernels in `CuDNNLSTM` and `CuDNNGRU` are still in wrong order. I think it would be cleaner to fix it in `_canonical_to_params()`.

**Comment by fchollet** [comment]
We won't do that. We cannot introduce a savefile-breaking change, even for a new layer that has only been recently introduced.

In general, `CuDNNLSTM` and `LSTM` are different layers, with a different number of weights, and there should be no expectation that loading the weights of one into another would work.

What we can do, is have a check and conversion step in the weight loading preprocessor in `topology`. If layer is `LSTM` and we get the number of weights corresponding to `CuDNNLSTM`, then check their shapes, and do the conversion.

This enables a model containing `CuDNNLSTM` to be converted to `LSTM` for inference and CPU, and have savefiles be loadable.

**Comment by myutwo150** [comment]
Okay, I'll change the code to do that.

File: keras/engine/topology.py [hunk 2]
```diff
@@ -3027,6 +3027,27 @@ def preprocess_weights_for_loading(layer, weights,
...
             if layer.__class__.__name__ == 'ConvLSTM2D':
                 weights[1] = np.transpose(weights[1], (3, 2, 0, 1))
+
+    # convert the weights of CuDNNLSTM so that they could be loaded into LSTM
+    if layer.__class__.__name__ == 'LSTM':
+        # determine if we're loading a CuDNNLSTM layer from the number of bias weights:
+        # CuDNNLSTM has (units * 8) weights; while LSTM has (units * 4)
+        units = weights[1].shape[0]
+        bias = weights[2]
+        if len(bias) == units * 8:
```
Comment (fchollet) [inline keras/engine/topology.py]: Please put this paragraph in the `if layer.__class__.__name__ == 'LSTM':` that already exists above

Comment (myutwo150) [reply keras/engine/topology.py]: I thought the `if layer.__class__.__name__ == 'LSTM':` that already exists above will only be executed if `original_keras_version == '1'`. This conversion is for weights saved under Keras>=2.0.8, isn't it?

Comment (fchollet) [reply keras/engine/topology.py]: You're right. Never mind then.

**Comment by charlesbensimon** [comment]
What about the opposite : Loading `LSTM` weights into `CuDNNLSTM` ?
As the `CuDNNLSTM` has no `dropout` or `recurrent_dropout`, in most cases training has to be done using an `LSTM`, and it would be great if we could use a `CuDNNLSTM` for inference

**Comment by bzamecnik** [comment]
It seems that the two GRU definitions come from confusion in subsequence version of the original GRU paper [Learning Phrase Representations using RNN Encoder-Decoder for Statistical Machine Translation](https://arxiv.org/abs/1406.1078).

In versions [v1](https://arxiv.org/pdf/1406.1078v1.pdf) and v2 they first projected previous state and then multiplied by reset gate:
1406.1078v1

However in [v3](https://arxiv.org/pdf/1406.1078v3.pdf) they changed the order and first gated the hidden state before projection:
image

Note that [On the Properties of Neural Machine Translation: Encoder–Decoder Approaches](https://arxiv.org/abs/1409.1259) and [Empirical Evaluation of Gated Recurrent Neural Networks on Sequence Modeling](https://arxiv.org/abs/1412.3555) and not the original definitions of GRU, although latter is focused on that. Also both of them follow the v3 definition.

CuDNN follows the original definition, probably since it allows merging kernels into a single GEMM, as beautifully described in [Optimizing RNNs with Differentiable Graphs](http://svail.github.io/diff_graphs/) - 5.2.3. Variant 3: Move the Reset Gate. It's also [consistent with LSTM](https://devblogs.nvidia.com/parallelforall/optimizing-recurrent-neural-networks-cudnn-5/). Unfortunately in v3 the authors changed the definition and that's what is implemented in TensorFlow and Keras.

Another problem is that the original paper doesn't specify biases. CuDNN use separate bias for each kernel, while Keras fuses both biases into one (which is reasonable given the v3 definition).

In [Optimizing RNNs with Differentiable Graphs](http://svail.github.io/diff_graphs/) they say that "While this changes the function being computed, it has similar functionality, and produces extremely similar training / accuracy performance in many of our experiments."

It's clear that it's necessary to maintain compatibility, ie. be able to load trained weights into the GRU layer. On the other hand CuDNN provides a significant speed-up and not being able to make inference on CPU is quite unfortunate. Since both architectures differ a tiny bit, there's a possibility of either making a copy of `keras.layers.GRU` and changing it to be compatible with CuDNN, or making `keras.layers.GRU` itself support both conventions, so that it can be used instead of `CuDNNGRU` when we want to infer on CPU. Anyway the v1 implementation would allow optimized implementation (merged GEMM even without CuDNN).

And by the way [TF is not afraid of converting biases](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py#L636) between both conventions there and back a hacky way - sum, divide to halves. :)

Any ideas about that? Thanks.

**Comment by bzamecnik** [comment]
I implemented a fork of GRU layer which is compatible with CuDNN and weight conversion and loading works, so that predictions are equals (up to numerical precision eps). So far it's in my git repo: https://github.com/bzamecnik/keras/commit/89534516848c3fdb4582ca0f1ebb998c78c4ea4e. So far the implementations are separate but I'd like to merge them if possible, adding just some argument to GRU layer, eg. variant = {'default', 'cudnn_compatible'} or {'reset_after', 'reset_before'}.

Is main Keras suitable for having two GRU conventions or should it go rather to keras-contrib?
